src/libraries/System.Private.CoreLib/src/System/Buffers/DiagnosticArrayPool.cs

```diff
@@ -0,0 +1,106 @@
+        /// it becomes collectible, the finalizable object's finalizer can try to validate that the
+        /// array wasn't written to post-return.
+        /// </remarks>
+        private static readonly ConditionalWeakTable<T[], ReturnedArrayState> s_returnedArrays = [];
+
+        /// <inheritdoc />
+        public override T[] Rent(int minimumLength)
+        {
+            ArgumentOutOfRangeException.ThrowIfLessThan(minimumLength, 0);
+
+            // Allocate an array that has a size that's a power-of-2 and at least 16, to match the behavior of ArrayPool.Shared.
+            minimumLength = Math.Max(minimumLength, 16);
+            minimumLength = (int)Math.Min(int.MaxValue, BitOperations.RoundUpToPowerOf2((uint)minimumLength));
+            var array = new T[minimumLength];
+
+            // For unmanaged Ts, fill the array with garbage to help catch bugs in consumers that might
+            // be assuming zero initialization.
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                Random.Shared.NextBytes(AsBytes(array));
+            }
+
+            return array;
+        }
+
+        /// <inheritdoc />
+        public override void Return(T[] array, bool clearArray = false)
+        {
+            ArgumentNullException.ThrowIfNull(array);
+
+            // Clear it regardless of clearArray so that we can easily check whether any non-zero writes
+            // were performed post-return. This could also help to catch bugs in consumers that might be
+            // returning arrays and then still using them for their contents.
+            Array.Clear(array);
+
+            // Locking isn't necessary for safety of data structures, but rather to serialize access
+            // to help catch concurrent misuse.
+            lock (s_returnedArrays)
+            {
+                if (s_returnedArrays.TryGetValue(array, out ReturnedArrayState? state))
+                {
+                    // Array was previously returned.
+                    Environment.FailFast(
+                        $"The array being returned to the ArrayPool was double-returned.{Environment.NewLine}" +
+                        $"Current stack: {Environment.StackTrace}{Environment.NewLine}" +
+                        $"Previous stack: {state.Stack}");
+                }
+
+                // Put the array into the CWT so that we can:
+                // 1) Detect if it's returned again.
+                // 2) Detect if it's modified after being returned.
+                s_returnedArrays.Add(array, new(array));
+            }
+        }
+
+        /// <summary>Gets a byte span over the array.</summary>
+        private static Span<byte> AsBytes(T[] array) =>
+            new Span<byte>(
+                ref Unsafe.As<T, byte>(ref MemoryMarshal.GetArrayDataReference(array)),
+                array.Length * Unsafe.SizeOf<T>());
```

Comment:
can this lead to overflows where current ArrayPool is fine?

Reply:
Yes, this needs to be fixed.